cli/graph.rs

```diff
@@ -0,0 +1,908 @@
+        .or_default();
+      if desc.kind == DependencyKind::ExportType
+        || desc.kind == DependencyKind::ImportType
+      {
+        dep.maybe_type = Some(specifier);
+      } else {
+        dep.maybe_code = Some(specifier);
+      }
+      if let Some(types_specifier) = maybe_types_specifier {
+        dep.maybe_type = Some(types_specifier);
+      }
+    }
+
+    self.maybe_parsed_module = Some(parsed_module);
+    Ok(())
+  }
+
+  fn resolve_import(
+    &self,
+    specifier: &str,
+    maybe_location: Option<Location>,
+  ) -> Result<ModuleSpecifier> {
+    let maybe_resolve = if let Some(import_map) = self.maybe_import_map.clone()
+    {
+      import_map
+        .borrow()
+        .resolve(specifier, self.specifier.as_str())?
+    } else {
+      None
+    };
+    let specifier = if let Some(module_specifier) = maybe_resolve {
+      module_specifier
+    } else {
+      ModuleSpecifier::resolve_import(specifier, self.specifier.as_str())?
+    };
+
+    let referrer_scheme = self.specifier.as_url().scheme();
+    let specifier_scheme = specifier.as_url().scheme();
+    let location = maybe_location.unwrap_or(Location {
+      filename: self.specifier.to_string(),
+      line: 0,
+      col: 0,
+    });
+
+    // Disallow downgrades from HTTPS to HTTP
+    if referrer_scheme == "https" && specifier_scheme == "http" {
+      return Err(InvalidDowngrade(specifier.clone(), location).into());
+    }
+
+    // Disallow a remote URL from trying to import a local URL
+    if (referrer_scheme == "https" || referrer_scheme == "http")
+      && !(specifier_scheme == "https" || specifier_scheme == "http")
+    {
+      return Err(InvalidLocalImport(specifier.clone(), location).into());
+    }
+
+    Ok(specifier)
+  }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Stats(Vec<(String, u128)>);
+
+impl<'de> Deserialize<'de> for Stats {
+  fn deserialize<D>(deserializer: D) -> result::Result<Self, D::Error>
+  where
+    D: Deserializer<'de>,
+  {
+    let items: Vec<(String, u128)> = Deserialize::deserialize(deserializer)?;
+    Ok(Stats(items))
+  }
+}
+
+impl fmt::Display for Stats {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    for (key, value) in self.0.clone() {
+      write!(f, "{}: {}", key, value)?;
+    }
+
+    Ok(())
+  }
+}
+
+/// A structure which provides options when transpiling modules.
+#[derive(Debug, Default)]
+pub struct TranspileOptions {
+  /// If `true` then debug logging will be output from the isolate.
+  pub debug: bool,
+  /// A string of configuration data that augments the the default configuration
+  /// passed to the TypeScript compiler.  This is typically the contents of a
+  /// user supplied `tsconfig.json`.
+  pub maybe_config: Option<String>,
+}
+
+/// The transpile options that are significant out of a user provided tsconfig
+/// file, that we want to parse out of the config.
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TranspileTsOptions {
```

Comment:
What's the difference between `TranspileOptions` and `TranspileTsOptions`?

Reply:
I guess I should make the clearer... `TranspileOptions` are options for the `.transpile()` method, where as `TranspileTsOptions` are just the compiler config options we care about when transpiling, which we can easily deserialize from the config JSON.  So maybe it is `TranspileConfigOptions` or `TranspileTsConfigOptions`?

`TranspileTsOptions` I don't think needs to be public as well.